src/Handler/UnionHandler.php

```diff
@@ -47,46 +48,90 @@ public function serializeUnion(
-    {
-        if ($data instanceof \SimpleXMLElement) {
-            throw new RuntimeException('XML deserialisation into union types is not supported yet.');
+            return $context->getNavigator()->accept($data, $resolvedType);
         }
-
-        return $this->matchSimpleType($data, $type, $context);
     }
 
-    private function matchSimpleType(mixed $data, array $type, Context $context)
+    public function deserializeUnion(DeserializationVisitorInterface $visitor, mixed $data, array $type, DeserializationContext $context): mixed
     {
-        $dataType = $this->determineType($data, $type, $context->getFormat());
-        $alternativeName = null;
-
-        if (isset(static::$aliases[$dataType])) {
-            $alternativeName = static::$aliases[$dataType];
+        if ($data instanceof \SimpleXMLElement) {
+            throw new RuntimeException('XML deserialisation into union types is not supported yet.');
         }
 
         foreach ($type['params'] as $possibleType) {
-            if ($possibleType['name'] === $dataType || $possibleType['name'] === $alternativeName) {
-                return $context->getNavigator()->accept($data, $possibleType);
+            $finalType = null;
+
+            if (!$context->getMetadataStack()->isEmpty()) {
+                $propertyMetadata = $context->getMetadataStack()->top();
+                if (null !== $propertyMetadata->unionDiscriminatorField) {
+                    if (!array_key_exists($propertyMetadata->unionDiscriminatorField, $data)) {
+                        throw new NonVisitableTypeException('Union Discriminator Field \'' . $propertyMetadata->unionDiscriminatorField . '\' not found in data');
+                    }
+
+                    $lkup = $data[$propertyMetadata->unionDiscriminatorField];
+                    if (!empty($propertyMetadata->unionDiscriminatorMap)) {
+                        if (array_key_exists($lkup, $propertyMetadata->unionDiscriminatorMap)) {
+                            $finalType = [
+                                'name' => $propertyMetadata->unionDiscriminatorMap[$lkup],
+                                'params' => [],
+                            ];
+                        } else {
+                            throw new NonVisitableTypeException('Union Discriminator Map does not contain key \'' . $lkup . '\'');
+                        }
+                    } else {
+                        $finalType = [
+                            'name' => $lkup,
+                            'params' => [],
+                        ];
+                    }
+                }
+            }
+
+            if (null !== $finalType && null !== $finalType['name']) {
+                return $context->getNavigator()->accept($data, $finalType);
+            } else {
+                foreach ($type['params'] as $possibleType) {
```

Comment:
something is odd with this `foreach` loops, you have here a nested loop the same variable (line 70 and 102). shouldn't they be independent?